### ISSUE TITLE
[Ray Serve] | Replicas keep failing health check under high QPS while scaling up 

### ISSUE THREAD

**Issue by danishshaikh556** [ray-project/ray#46702]
### What happened + What you expected to happen

Ray Scaling up or Deploys when having high traffic cause lots of replica health check failures

![Screenshot 2024-06-26 at 3 42 00 PM](https://github.com/user-attachments/assets/cf6f1deb-e452-4a5a-bd06-0740d52e8560)
![Screenshot 2024-06-26 at 3 39 51 PM](https://github.com/user-attachments/assets/151f2677-d9e8-4dfb-aa9a-cd89be1bea4e)
![Screenshot 2024-06-26 at 3 39 15 PM](https://github.com/user-attachments/assets/d3d5d0a6-478d-4d86-b94a-25cdf354fb42)
[serve_controller_152 (1).log](https://github.com/user-attachments/files/16289892/serve_controller_152.1.log)

### Versions / Dependencies

Ray 2.20.0
Kuberay 1.0.0

### Reproduction script

N/A

### Issue Severity

High: It blocks me from completing my task.

**Comment by danishshaikh556** [comment]
Actors when failing usually have this error
```
ERROR 2024-07-18 16:15:15,862 TextSearchRerankingTransformerV10_TFServe j28b73mh af31c081-2e1e-4658-9cc8-7f95acf3ec8e /serve replica.py:359 - Request failed:
[36mray::ServeReplica:TextSearchRerankingTransformerV10:TFServe.handle_request_with_rejection()[39m (pid=23474, ip=100.82.149.124)
  File "/home/ray/anaconda3/lib/python3.9/site-packages/ray/serve/_private/utils.py", line 168, in wrap_to_ray_error
    raise exception
  File "/home/ray/anaconda3/lib/python3.9/site-packages/ray/serve/_private/replica.py", line 1132, in call_user_method
    await self._call_func_or_gen(
  File "/home/ray/anaconda3/lib/python3.9/site-packages/ray/serve/_private/replica.py", line 856, in _call_func_or_gen
    result = await result
  File "/tmp/ray/session_2024-07-15_15-46-35_769995_8/runtime_resources/working_dir_files/s3_sr-search-data-886239521314_sagemaker_deploy_text_search_reranking_trainer_v10_dag-ray_2024-07-15_model/serve.py", line 392, in __call__
    input_data: dict = await starlette_request.json()
  File "/home/ray/anaconda3/lib/python3.9/site-packages/starlette/requests.py", line 251, in json
    body = await self.body()
  File "/home/ray/anaconda3/lib/python3.9/site-packages/starlette/requests.py", line 244, in body
    async for chunk in self.stream():
  File "/home/ray/anaconda3/lib/python3.9/site-packages/starlette/requests.py", line 238, in stream
    raise ClientDisconnect()
starlette.requests.ClientDisconnect
```

**Comment by ravishtiwari** [comment]
I have seen this happening in high QPS scenarios, with similar warning:

```
WARNING 2024-06-26 14:58:55,964 controller 537 deployment_state.py:775 - Health check for Replica...  failed: 
```

**Comment by zcin** [comment]
@danishshaikh556 Is my understanding correct:
When the deployment upscales and brings up new replicas, most (or all?) requests that are sent to the replica fail. Then you have a custom health check for this deployment that checks the success rate of requests, and that health check fails because of all the request failures. This causes the controller to kill the replicas, which then causes the p99 latency to spike?

**Comment by danishshaikh556** [comment]
Yes this is correct with only one correlation, the replicas failing aren’t
necessarily just the newly spun up ones, old ones mostly start failing too.

Thanks

Danish Shaikh
LinkedIn <https://www.linkedin.com/in/danishshaikh556/> | 530-680-4102

**Comment by danishshaikh556** [comment]
@zcin adding more snapshot of metrics of what our system looks like when this happends
![Screenshot 2024-07-24 at 4 59 36 PM](https://github.com/user-attachments/assets/ba956422-1cf3-4590-bd57-468130021554)
If you look you can see on high qps when scale up is happening replicas are dying, thus adding to latency(See latency replica graph)

**Comment by sridif** [comment]
had a similar issue that got solved by removing the initial_replicas in the autoscaling config.

**Comment by sridif** [comment]
perhaps our case is different - but leaving a comment here incase it helps ... 

we faced a similar issue when we had latency in one our deployment - that lead to the deployment scaling up the number of replicas as per our autoscaling config. But after few seconds we saw both the original replicas and the scaled up replicas ending up in error state - and getting deleted. 

we saw in the serve controller logs that the number of replicas was higher that initial_relicas. That gave us a clue that some how the deployment thinks its in a initial state and kills all the running replicas.

after removing the initial_replicas from our autoscaling config and relying solely on min_replicas we were able to aviod this issue.
